### PR TITLE
Support for injecting callbacks before the value is returned to the client

### PR DESCRIPTION
**Pull request by ian-h-chamberlain**
This is a pretty hefty PR, but it includes the base work done in [this repo](https://github.com/billr1/python-netsnmpagent) to add a few more types and the basic idea of callbacks.

I've made some changes to refactor those changes and simplified the examples a bit in an effort to make this presentable upstream. Manually testing all of the included examples and integration tests resulted in success for both python2.7 and python3.6, so as far as I can tell this should be good to go. I'll address any comments here if needed, then submit a PR upstream if it looks good.

As a note, `examples/run_callback_agent.sh` is almost completely copied from `examples/run_simple_agent.sh` and `examples/callback_agent.py` is pretty much the same as `examples/simple_agent.py` except for the use of a class to encapsulate the updating of the registered instances.

### CODE REVIEW

File: netsnmpagent.py [hunk 1]
```diff
@@ -482,20 +527,25 @@ def __init__(self):
...
 					val = self._cvar.value
 
-					if isnum(val):
-						# Python 2.x will automatically switch from the "int"
-						# type to the "long" type, if necessary. Python 3.x
-						# has no limits on the "int" type anymore.
-						val = int(val)
-					else:
-						val = u(val)
+					if self._asntype != ASN_OPAQUE_FLOAT and self._asntype != ASN_OPAQUE_DOUBLE:
```
Comment (apt-itude) [inline netsnmpagent.py:541]: You could do `if self._asntype not in {ASN_OPAQUE_FLOAT, ASN_OPAQUE_DOUBLE}`

File: netsnmpagent.py [hunk 2]
```diff
@@ -669,7 +760,76 @@ def update(self, val):
...
+					)
+					if result != 0:
+						raise netsnmpAgentException("Error registering variable with net-snmp!")
+
+					if self._callback_handler is not None:
+						_inject_custom_handler(self._callback_handler, handler_reginfo)
+
+					# Finally, we keep track of all registered SNMP objects for the
+					# getRegistered() method.
+					agent._objs[context][oidstr] = self
```
Comment (apt-itude) [inline netsnmpagent.py:813]: Is this guaranteed not to raise `KeyError` on the `context` lookup? If not, there should probably be some error handling.

**Review by ecarson — APPROVED** [review]

File: netsnmpagent.py [hunk 3]
```diff
@@ -669,7 +760,76 @@ def update(self, val):
...
+
+			def value(self):
+				# Get boolean representation of TruthValue.
+				return True if self._cvar.value == TV_TRUE else False
```
Comment (apt-itude) [inline netsnmpagent.py:817]: You could just return `self._cvar.value == TV_TRUE`

**Review by apt-itude — APPROVED** [review]
No real blockers, looks good!

File: netsnmpagent.py (hunk 2 above)
Comment (ian-h-chamberlain) [reply netsnmpagent.py:813]: I'm not sure – this pattern seems to be common to all of the various types.

Actually all the new type implementations in this PR are not from my commits – I pulled them from the upstream pull request [here](https://github.com/pief/python-netsnmpagent/pull/35). Could probably stand to be cleaned up a bit (less copy-paste etc) but I tried to keep my changes as low touch as possible.

**Comment by ian-h-chamberlain** [comment]
Merging so we can easily publish a package to the internal PyPI server. If we end up needing further changes I can do some cleaning up / more refactoring at that point.

**Pull request merged**